privacyidea/lib/policydecorators.py

```diff
@@ -156,28 +157,37 @@ def auth_cache(wrapped_function, user_object, passw, options=None):
     options = options or {}
     g = options.get("g")
     auth_cache_dict = None
+
     if g:
         auth_cache_dict = Match.user(g, scope=SCOPE.AUTH, action=ACTION.AUTH_CACHE,
                                      user_object=user_object).action_values(unique=True, write_to_audit_log=False)
         if auth_cache_dict:
-            # verify in cache and return an early success
             auth_times = list(auth_cache_dict)[0].split("/")
+
             # determine first_auth from policy!
             first_offset = parse_timedelta(auth_times[0])
+            first_auth = datetime.datetime.utcnow() - first_offset
+            last_auth = first_auth  # Default if no last auth exists
+            max_auths = 0  # Default value, 0 has no effect on verification
 
+            # Use auth cache when number of allowed authentications is defined
             if len(auth_times) == 2:
-                # Determine last_auth from policy
-                last_offset = parse_timedelta(auth_times[1])
+                if re.match(r"^\d+$", auth_times[1]):
+                    max_auths = int(auth_times[1])
```

Comment:
```suggestion
                    max_auths = int(auth_times[1])
                    last_auth = None
```

Like this? I am not sure about what difference it actually makes here.

Reply:
Now this is a tricky one :-)
Actually there are several things going on here:
- You used `models.AuthCache.__init__()` to create entries in the authcache table for the tests, unfortunately this doesn't create a `last_auth` entry (`None` by default), so basically this is an incomplete entry (You need to have a `last_auth` entry, otherwise it shouldn't exist in the authcache table).
- When rewriting this functionality we set `last_auth = first_auth` as a default, this was passed to `verify_in_cache()` which in turn created a db-query from that. But since the `last_auth` entry wasn't set in the db, the query returned nothing and the default authentication took place (this lead to a failed test in `test_lib_policydecorator.py:647`).

1. Without a valid `last_auth` in the authcache table the entry should be treated as invalid.
2. We should be able to pass the `last_auth` in any case to `verify_in_cache()`, if `last_auth = first_auth` this should yield all matching entries as if `last_auth=None` was passed.
3. Please delete the `last_auth = None` here and add a `last_auth` parameter in `test_lib_policydecorator.py:641`

Reply:
This should be fixed now too.